dm/pkg/checker/table_structure.go

```diff
@@ -79,68 +95,106 @@ func (c *TablesChecker) Check(ctx context.Context) *Result {
+	defer cancel()
+
+	checkFunc := func() {
+		defer checkWg.Done()
+		for {
+			select {
+			case <-checkCtx.Done():
+				return
+			case checkItem, ok := <-inCh:
+				if !ok {
+					return
+				}
+				table := checkItem.table
+				db := c.dbs[checkItem.sourceID]
+				statement, err := dbutil.GetCreateTableSQL(ctx, db, table.Schema, table.Name)
+				if err != nil {
+					// continue if table was deleted when checking
+					if isMySQLError(err, mysql.ErrNoSuchTable) {
+						continue
+					}
+					reMu.Lock()
+					markCheckError(r, err)
+					reMu.Unlock()
+					cancel()
+					break
 				}
-				markCheckError(r, err)
-				return r
-			}
 
-			opts := c.checkCreateSQL(ctx, statement)
-			if len(opts) > 0 {
-				options[tableName] = opts
-				statements[tableName] = statement
+				opts := c.checkCreateSQL(ctx, db, statement)
+				for _, opt := range opts {
+					opt.tableID = table.String()
+					tableMsg := "table " + opt.tableID + " "
+					reMu.Lock()
+					switch opt.state {
+					case StateWarning:
+						if r.State != StateFailure {
+							r.State = StateWarning
+						}
+						e := NewError(tableMsg + opt.errMessage)
+						e.Severity = StateWarning
+						e.Instruction = opt.instruction
+						r.Errors = append(r.Errors, e)
+					case StateFailure:
+						r.State = StateFailure
+						e := NewError(tableMsg + opt.errMessage)
+						e.Instruction = opt.instruction
+						r.Errors = append(r.Errors, e)
+					}
+					reMu.Unlock()
+				}
 			}
 		}
 	}
 
-	for name, opts := range options {
-		if len(opts) == 0 {
-			continue
+	concurrency := maxThreadNum
+	for instance := range c.tableMap {
+		db, ok := c.dbs[instance]
+		if !ok {
+			markCheckError(r, errors.NotFoundf("client for instance %s", instance))
+			return r
 		}
-		tableMsg := "table " + name + " "
+		maxConnections, err := utils.GetMaxConnections(ctx, db)
+		if err != nil {
+			markCheckError(r, err)
+			return r
+		}
+		concurrency = int(math.Min(float64(concurrency), float64(maxConnections/2)))
+	}
 
-		for _, option := range opts {
-			switch option.state {
-			case StateWarning:
-				if r.State != StateFailure {
-					r.State = StateWarning
-				}
-				e := NewError(tableMsg + option.errMessage)
-				e.Severity = StateWarning
-				e.Instruction = option.instruction
-				r.Errors = append(r.Errors, e)
-			case StateFailure:
-				r.State = StateFailure
-				e := NewError(tableMsg + option.errMessage)
-				e.Instruction = option.instruction
-				r.Errors = append(r.Errors, e)
+	for i := 0; i < concurrency; i++ {
+		checkWg.Add(1)
+		go checkFunc()
+	}
+
+	for sourceID, tables := range c.tableMap {
+		for _, table := range tables {
+			isDone := false
+			select {
+			case inCh <- &checkItem{table, sourceID}:
+			case <-checkCtx.Done():
+				isDone = true
+			}
+			if isDone {
+				break
 			}
 		}
 	}
 
+	close(inCh)
+	checkWg.Wait()
+	log.L().Logger.Info("check table structure over", zap.Time("start time", startTime), zap.String("speed time", time.Since(startTime).String()))
```

Comment:
If there are none tables checked due to ctx.Done, we should check and adjust the check result to Canceled if it is still succeeded.